src/matrix/room/timeline/Timeline.js

```diff
@@ -236,7 +254,111 @@ export class Timeline {
     /** @package */
     addEntries(newEntries) {
         this._addLocalRelationsToNewRemoteEntries(newEntries);
+        this._updateEntriesFetchedFromHomeserver(newEntries);
+        this._moveEntryToRemoteEntries(newEntries);
         this._remoteEntries.setManySorted(newEntries);
+        this._loadContextEntriesWhereNeeded(newEntries);
+    }
+
+    /**
+     * Update entries based on newly received events.
+     * This is specific to events that are not in the timeline but had to be fetched from the homeserver.
+     */
+    _updateEntriesFetchedFromHomeserver(entries) {
+        /**
+         * Updates for entries in timeline is handled by remoteEntries observable collection
+         * Updates for entries not in timeline but fetched from storage is handled in this.replaceEntries()
+         * This code is specific to entries fetched from HomeServer i.e NonPersistedEventEntry
+         */
+        for (const entry of entries) {
+            const relatedEntry = this._contextEntriesNotInTimeline.get(entry.relatedEventId);
+            if (relatedEntry?.addRelation(entry)) {
+                // update other entries for which this entry is a context entry
+                relatedEntry.contextForEntries?.forEach(e => {
+                    this._remoteEntries.findAndUpdate(te => te.id === e.id, () => { return { reply: relatedEntry }; });
+                });
+            }
+        }
+    }
+
+    /**
+     * If an event we had to fetch from hs/storage is now in the timeline (for eg, due to gap fill),
+     * remove the event from _contextEntriesNotInTimeline since it is now in remoteEntries
+     */
+    _moveEntryToRemoteEntries(entries) {
+        for (const entry of entries) {
+            const fetchedEntry = this._contextEntriesNotInTimeline.get(entry.id);
+            if (fetchedEntry) {
+                fetchedEntry.contextForEntries.forEach(e => e.setContextEntry(entry));
+                entry.updateFrom(fetchedEntry);
```

Comment:
Why do we need this? `contextForEntries` has been set by the previous line, not sure what else needs to be copied, as the entry that just came from the server has the latest event?

Also, to be completely consistent, we could also emit an update here for all the entries that have this entry as context? It likely wouldn't change anything on screen but it would ensure that the new entry is used everywhere down in the TilesCollection and the view, ...

Reply:
Previously, `updateFrom` was needed to copy over `contextFor` to the new entry. Since we now do that implicitly in `setContextEntry()`, we can remove it.